src/app/modules/auth/auth.middlewares.ts

```diff
@@ -92,3 +95,45 @@ export const validateVerifyOtpParams = celebrate({
       .message('Please enter a valid OTP'),
   }),
 })
+
+/**
+ * Middleware that only allows users with a valid bearer token to pass through to the next handler
+ */
+export const authenticateApiKey: ControllerHandler<
+  unknown,
+  unknown,
+  ApiReqBody
+> = (req, res, next) => {
+  const authorizationHeader = req.headers.authorization
+  if (!authorizationHeader) {
+    return res
+      .status(StatusCodes.UNAUTHORIZED)
+      .json({ message: 'Authorisation header is missing' })
+  }
+  const [bearerString, apiKey] = authorizationHeader.split(BEARER_SEPARATOR)
+  if (bearerString !== BEARER_STRING) {
+    return res
+      .status(StatusCodes.BAD_REQUEST)
+      .json({ message: 'Invalid authorisation header format' })
+  }
+  const [apiEnv, apiVersion, userId, token] = apiKey.split(API_KEY_SEPARATOR)
+  if (!apiEnv || !apiVersion || !userId || !token) {
+    return res
+      .status(StatusCodes.BAD_REQUEST)
+      .json({ message: 'Invalid API key format' })
+  }
```

Comment:
A regex validation scheme here could be more precise in validating the shape of the header and the key 🤔 . I can help craft that.

Reply:
Taking a stab...

```js
new RegExp(
  '^' + // ensure start of string
    '(?:Bearer)' + // Non-capturing group: marching the "Bearer" word
    '[ ]' + // Single space
    "(?'apiEnvToken'w+)" + // Capture group {apiEnvToken}
    '_' + // delimiter
    "(?'apiVerToken'vd+)" + // Capture group {apiVerToken}: matching starting with character v, and at least 1 digit
    '_' + // delimiter
    "(?'userIdToken'[[:alnum:]]{24})" + // Capture group {userIdToken}: matching a mongodb objectid, 24 characters of alphanumeric
    '_' + // delimiter
    "'apiToken'(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}={2}))" + // Capture group {userIdToken}: matching a mod4 base64 value
    '$', // ensure end of string,
)```

Reply:
Nice @KenLSM !

I don't think we need to go to that extent of validation though 😅. The only thing we really care about is to extract `userId`. I added in a simplified one for now.

Just to note also: bcrypt base64 encoding uses `/.` rather than `/+` (see [doc](https://en.wikipedia.org/wiki/Base64#Applications_not_compatible_with_RFC_4648_Base64))